File: app/src/main/java/com/apache/fastandroid/topic/support/imageloader/GlideImageLoader.java
package com.apache.fastandroid.topic.support.imageloader;

import android.content.Context;
import android.content.res.Resources;
import android.graphics.Bitmap;
import android.os.Looper;
import android.text.TextUtils;
import android.util.Log;
import android.view.View;
import android.widget.ImageView;

import com.bumptech.glide.DrawableTypeRequest;
import com.bumptech.glide.GenericRequestBuilder;
import com.bumptech.glide.GifRequestBuilder;
import com.bumptech.glide.Glide;
import com.bumptech.glide.RequestManager;
import com.bumptech.glide.load.engine.DiskCacheStrategy;
import com.bumptech.glide.load.resource.gif.GifDrawable;
import com.bumptech.glide.request.FutureTarget;
import com.bumptech.glide.request.animation.GlideAnimation;
import com.bumptech.glide.request.target.SimpleTarget;
import com.bumptech.glide.request.target.Target;
import com.tesla.framework.component.imageloader.IImageLoaderstrategy;
import com.tesla.framework.component.imageloader.IImagePahtFromCache;
import com.tesla.framework.component.imageloader.ImageLoaderOptions;

import java.io.File;
import java.util.concurrent.ExecutionException;


/**
 * Created by Administrator on 2017/3/21 0021.
 */

public class GlideImageLoader implements IImageLoaderstrategy,IImagePahtFromCache {

    @Override
    public void showImage(ImageLoaderOptions options) {
        GenericRequestBuilder mGenericRequestBuilder = init(options);
        if (mGenericRequestBuilder != null) {
            showImageLast(mGenericRequestBuilder, options);
        }
    }


    @Override
    public void cleanMemory(Context context) {
        if (Looper.myLooper() == Looper.getMainLooper()) {
            Glide.get(context).clearMemory();
        }
    }




    public DrawableTypeRequest getGenericRequestBuilder(RequestManager manager, ImageLoaderOptions options) {

        if (!TextUtils.isEmpty(options.getUrl())) {
            return manager.load(options.getUrl());
        }
        return manager.load(options.getResource());
    }

    public GenericRequestBuilder init(ImageLoaderOptions options) {
        View v = options.getViewContainer();

        //存在问题
        // java.lang.IllegalArgumentException You cannot launch a load for a destroyed activity
        //RequestManager manager=getRequestManager(v.getContext());
        RequestManager manager = Glide.with(options.getContext());
        if (v instanceof ImageView) {
            GenericRequestBuilder mDrawableTypeRequest = getGenericRequestBuilder(manager, options).asBitmap();
            //装载参数
            mDrawableTypeRequest = loadGenericParams(mDrawableTypeRequest, options);
            return mDrawableTypeRequest;
        }
        return null;
    }


    private GenericRequestBuilder loadGenericParams(GenericRequestBuilder mGenericRequestBuilder, final ImageLoaderOptions options) {
        final View view = options.getViewContainer();
        GenericRequestBuilder builder = mGenericRequestBuilder;
        if (mGenericRequestBuilder instanceof DrawableTypeRequest) {
            if (options.isCrossFade()) {
                ((DrawableTypeRequest) mGenericRequestBuilder).crossFade();
            }
            if (options.isAsGif()) {
                builder = ((DrawableTypeRequest) mGenericRequestBuilder).asGif();
            }
        }
        builder.skipMemoryCache(options.isSkipMemoryCache());
        if (options.getImageSize() != null) {
            int width = getSize(options.getImageSize().getWidth(), view);
            int height = getSize(options.getImageSize().getHeight(), view);
            Log.i("tag ","load params " + options.getImageSize().getWidth() + "  : " + options.getImageSize().getHeight());
            builder.override(width, height);
        }
        if (options.getHolderId() > 0) {
            builder.placeholder(options.getHolderId());
        }

        if (options.getHolderDrawable() != null){
            builder.placeholder(options.getHolderDrawable());
        }

        if (options.getErrorId() > 0) {
            builder.error(options.getErrorId());
        }

        if (options.getErrorDrawable() != null){
            builder.placeholder(options.getErrorDrawable());
        }



        if (options.getDiskCacheStrategy() != ImageLoaderOptions.DiskCacheStrategy.DEFAULT) {
            switch (options.getDiskCacheStrategy()) {
                case NONE:
                    builder.diskCacheStrategy(DiskCacheStrategy.NONE);
                    break;
                case All:
                    builder.diskCacheStrategy(DiskCacheStrategy.ALL);
                    break;
                case SOURCE:
                    builder.diskCacheStrategy(DiskCacheStrategy.SOURCE);
                    break;
                case RESULT:
                    builder.diskCacheStrategy(DiskCacheStrategy.RESULT);
                    break;
                default:
                    break;
            }
        }


        return builder;

    }

    private void showImageLast(GenericRequestBuilder mDrawableTypeRequest, ImageLoaderOptions options) {
        final ImageView img = (ImageView) options.getViewContainer();

        // 是否使用高斯模糊
        if (options.isBlurImage()) {
            // 具体的高斯模糊这里就不实现了，直接展示图片
            mDrawableTypeRequest.into(img);
            return;
        }
        // 是否展示一个gif
        if (options.isAsGif()) {
            ((GifRequestBuilder) mDrawableTypeRequest).dontAnimate().into(new SimpleTarget<GifDrawable>() {
                @Override
                public void onResourceReady(GifDrawable resource, GlideAnimation<? super GifDrawable> glideAnimation) {
                    img.setImageDrawable(resource);
                    resource.start();
                }
            });
            return;
        }
//        if (options.getTarget() != null) {
//            mDrawableTypeRequest.into(options.getTarget());
//            return;
//        }
        mDrawableTypeRequest.into(img);
    }

    /**
     * 获取资源尺寸
     *
     * @param resSize
     * @return 默认返回原始尺寸
     */
    private int getSize(int resSize, View container) {
        if (resSize <= 0) {
            return SimpleTarget.SIZE_ORIGINAL;
        } else {
            try {
                return container.getContext().getResources().getDimensionPixelOffset(resSize);

            } catch (Resources.NotFoundException e) {
                e.printStackTrace();
                Log.e("","I got !!!");
                return resSize;
            }
        }
    }

    private Context mContext;
    @Override
    public void init(Context context) {
        mContext = context;
    }

    @Override
    public void pause(Context context) {
        if (Looper.myLooper() == Looper.getMainLooper()){
            Glide.with(context).pauseRequests();
        }
    }

    @Override
    public void resume(Context context) {
        if (Looper.myLooper() == Looper.getMainLooper()){
            Glide.with(context).resumeRequests();
        }
    }

    @Override
    public String getImagePahtFromCache(String url) {
        FutureTarget<File>  futuer = Glide.with(mContext).load(url).downloadOnly(Target.SIZE_ORIGINAL,Target.SIZE_ORIGINAL);
        try {
            File cacheFile = futuer.get();
            return cacheFile.getAbsolutePath();
        } catch (InterruptedException e) {
            e.printStackTrace();
        } catch (ExecutionException e) {
            e.printStackTrace();
        }
        return null;
    }

    @Override
    public Bitmap getBitmapFromCache(String url) {
        Bitmap myBitmap = null;
        try {
            myBitmap = Glide.with(mContext)
                    .load(url)
                    .asBitmap() //必须
                    .centerCrop()
                    .into(Target.SIZE_ORIGINAL, Target.SIZE_ORIGINAL)
                    .get();
        } catch (InterruptedException e) {
            e.printStackTrace();
        } catch (ExecutionException e) {
            e.printStackTrace();
        }
        return myBitmap;
    }
}

File: fastFramework/src/main/java/com/tesla/framework/component/imageloader/GlideImageLoader.kt
package com.tesla.framework.component.imageloader

import android.content.Context
import android.content.res.Resources
import android.graphics.Bitmap
import android.os.Looper
import android.text.TextUtils
import android.util.Log
import android.view.View
import android.widget.ImageView
import com.bumptech.glide.*
import com.bumptech.glide.load.engine.DiskCacheStrategy
import com.bumptech.glide.load.resource.gif.GifDrawable
import com.bumptech.glide.request.animation.GlideAnimation
import com.bumptech.glide.request.target.SimpleTarget
import com.bumptech.glide.request.target.Target
import java.util.concurrent.ExecutionException
import kotlin.Int
import kotlin.String

/**
 * Created by Administrator on 2017/3/21 0021.
 */
class GlideImageLoader : IImageLoaderstrategy, IImagePahtFromCache {
    override fun showImage(imageView: ImageView, url: String?, options: ImageOptions?) {
        val mGenericRequestBuilder = init(imageView, url, options)
        showImageLast(imageView, mGenericRequestBuilder, options)
    }

    override fun cleanMemory(context: Context) {
        if (Looper.myLooper() == Looper.getMainLooper()) {
            Glide.get(context).clearMemory()
        }
    }

    private fun getGenericRequestBuilder(
        manager: RequestManager,
        url: String?,
        options: ImageOptions?
    ): DrawableTypeRequest<*> {
        return if (!TextUtils.isEmpty(url)) {
            manager.load(url)
        } else manager.load(options!!.holderResId)
    }

    fun init(
        v: ImageView,
        url: String?,
        options: ImageOptions?
    ): GenericRequestBuilder<*, *, *, *> {
        val manager = Glide.with(v.context)
        var mDrawableTypeRequest: GenericRequestBuilder<*, *, *, *> =
            getGenericRequestBuilder(manager, url, options).asBitmap()
        //装载参数
        mDrawableTypeRequest = loadGenericParams(v, mDrawableTypeRequest, options)
        return mDrawableTypeRequest
    }

    private fun loadGenericParams(
        view: ImageView?,
        mGenericRequestBuilder: GenericRequestBuilder<*, *, *, *>,
        options: ImageOptions?
    ): GenericRequestBuilder<*, *, *, *> {
        var builder = mGenericRequestBuilder
        if (mGenericRequestBuilder is DrawableTypeRequest<*>) {
            if (options!!.isCrossFade!!) {
                mGenericRequestBuilder.crossFade()
            }
            if (options.asGif) {
                builder = mGenericRequestBuilder.asGif()
            }
        }
        builder.skipMemoryCache(options!!.skipMemoryCache!!)
        if (options.imageSize != null) {
            val width = getSize(options.imageSize.width, view)
            val height = getSize(options.imageSize.height, view)
            Log.i(
                "tag ",
                "load params " + options.imageSize.width + "  : " + options.imageSize.height
            )
            builder.override(width, height)
        }
        if (options.holderResId != null) {
            builder.placeholder(options.holderResId)
        }
        if (options.holderDrawable != null) {
            builder.placeholder(options.holderDrawable)
        }
        if (null != options.errorResId) {
            builder.error(options.errorResId)
        }
        if (options.errorDrawable != null) {
            builder.placeholder(options.errorDrawable)
        }
        if (options.mDiskCacheStrategy != null && options.mDiskCacheStrategy !== ImageOptions.DiskCacheStrategy.DEFAULT) {
            when (options.mDiskCacheStrategy) {
                ImageOptions.DiskCacheStrategy.NONE -> builder.diskCacheStrategy(
                    DiskCacheStrategy.NONE
                )
                ImageOptions.DiskCacheStrategy.All -> builder.diskCacheStrategy(DiskCacheStrategy.ALL)
                ImageOptions.DiskCacheStrategy.SOURCE -> builder.diskCacheStrategy(
                    DiskCacheStrategy.SOURCE
                )
                ImageOptions.DiskCacheStrategy.RESULT -> builder.diskCacheStrategy(
                    DiskCacheStrategy.RESULT
                )
                else -> {}
            }
        }
        return builder
    }

    private fun showImageLast(
        imageView: ImageView,
        mDrawableTypeRequest: GenericRequestBuilder<*, *, *, *>,
        options: ImageOptions?
    ) {
//        final ImageView img = (ImageView) options.getViewContainer();

        // 是否使用高斯模糊
        if (options?.blurImage == true) {
            // 具体的高斯模糊这里就不实现了，直接展示图片
            mDrawableTypeRequest.into(imageView)
            return
        }
        // 是否展示一个gif
        if (options?.asGif == true) {
            (mDrawableTypeRequest as GifRequestBuilder<*>).dontAnimate()
                .into(object : SimpleTarget<GifDrawable?>() {


                    override fun onResourceReady(
                        resource: GifDrawable?,
                        glideAnimation: GlideAnimation<in GifDrawable?>?
                    ) {
                        imageView.setImageDrawable(resource)
                        resource?.start()
                    }
                })
            return
        }

        mDrawableTypeRequest.into(imageView)
    }

    /**
     * 获取资源尺寸
     *
     * @param resSize
     * @return 默认返回原始尺寸
     */
    private fun getSize(resSize: Int, container: View?): Int {
        return if (resSize <= 0) {
            SimpleTarget.SIZE_ORIGINAL
        } else {
            try {
                container!!.context.resources.getDimensionPixelOffset(resSize)
            } catch (e: Resources.NotFoundException) {
                e.printStackTrace()
                Log.e("", "I got !!!")
                resSize
            }
        }
    }

    private var mContext: Context? = null
    override fun init(context: Context) {
        mContext = context
    }

    override fun pause(context: Context) {
        if (Looper.myLooper() == Looper.getMainLooper()) {
            Glide.with(context).pauseRequests()
        }
    }

    override fun resume(context: Context) {
        if (Looper.myLooper() == Looper.getMainLooper()) {
            Glide.with(context).resumeRequests()
        }
    }

    override fun getImagePahtFromCache(url: String?): String? {
        val futuer =
            Glide.with(mContext).load(url).downloadOnly(Target.SIZE_ORIGINAL, Target.SIZE_ORIGINAL)
        try {
            val cacheFile = futuer.get()
            return cacheFile.absolutePath
        } catch (e: InterruptedException) {
            e.printStackTrace()
        } catch (e: ExecutionException) {
            e.printStackTrace()
        }
        return null
    }

    override fun getBitmapFromCache(url: String?): Bitmap? {
        var myBitmap: Bitmap? = null
        try {
            myBitmap = Glide.with(mContext)
                .load(url)
                .asBitmap() //必须
                .centerCrop()
                .into(Target.SIZE_ORIGINAL, Target.SIZE_ORIGINAL)
                .get()
        } catch (e: InterruptedException) {
            e.printStackTrace()
        } catch (e: ExecutionException) {
            e.printStackTrace()
        }
        return myBitmap
    }
}